src/box/memtx_tx.c

```diff
@@ -1835,8 +1839,17 @@ memtx_tx_tuple_clarify_slow(struct txn *txn, struct space *space,
 		if (story == NULL)
 			break;
 	}
-	if (!own_change && story != NULL)
-		memtx_tx_track_read_story(txn, space, story);
+	if (!own_change && story != NULL) {
+		/*
+		 * If the result tuple exists (is visible) - it is visible in
+		 * every index. But if we found a story of deleted tuple - we
+		 * should record that only in the given index this transaction
+		 * have found nothing by this key.
+		 */
+		int shift = index->dense_id & 63;
+		uint64_t mask = result == NULL ? 1ull << shift : UINT64_MAX;
+		memtx_tx_track_read_story(txn, space, story, mask);
```

Comment:
I would add 'panic' here in case 'memtx_tx_track_read_story' fails because of out of memory.

Reply:
Actually I don't know what to do in case of that error. Iterators don't have errors now. We cannot just stop iteration - it would be quite unexpected.